third_party/nvfuser/csrc/fusion_segmenter.cpp

```diff
@@ -811,82 +844,88 @@ std::vector<SegmentedEdge*> SegmentedFusion::castInputOutputToLowerPrecision(
-          std::inserter(merged_group_exprs, merged_group_exprs.end()));
+      edges_to_merged_groups[edge_tv].push_back(edge);
+    } else {
+      bundled_edges.push_back({edge});
     }
+  }
 
-    // Gather exprs that should be modified. Start with all use
-    // exprs.
-    auto uses_to_modify = edge_tv->uses();
-
-    const bool is_from_merged_groups =
-        std::find(groups_to_merge.begin(), groups_to_merge.end(), edge->from) !=
-        groups_to_merge.end();
+  for (const auto val : vals_of_edges_to_merged_groups) {
+    bundled_edges.emplace_back(edges_to_merged_groups.at(val));
+  }
 
-    // Uses of the edge value within the from group should not be replaced. This
-    // will cause the group to have an intermediate tensor
-    // tv -> float2half -> output
-    //            \ -> half2float -> other uses in group
-    // The conversion back and forth from half precision can hurt numerics.
-    // Collect expressions that use the edge value of concern within the from
-    // group to avoid replacing with the cast tensor.
-
-    // All expressions in the from group of the edge. When
-    // the from group is included in groups_to_merge, grab all exprs
-    // in the all groups_to_merge groups
-    std::unordered_set<Expr*> from_group_exprs = is_from_merged_groups
-        ? merged_group_exprs
-        : std::unordered_set<Expr*>{
-              edge->from->exprs().begin(), edge->from->exprs().end()};
+  // Go through all edges of the segmented fusion.
+  for (const auto& edges : bundled_edges) {
+    auto edge_tv = edges.at(0)->val->as<TensorView>();
 
-    uses_to_modify.erase(
-        std::remove_if(
-            uses_to_modify.begin(),
-            uses_to_modify.end(),
-            [&](Expr* edge_val_use_expr) {
-              return from_group_exprs.count(edge_val_use_expr);
-            }),
-        uses_to_modify.end());
+    // Gather exprs that should be modified. Start with all use
+    // exprs.
+    std::vector<Expr*> uses_to_modify;
 
-    // When this edge is an edge to the groups_to_merge groups, the
-    // associated TV will be cast to lower precision. Only modify the
-    // uses of tv contained within the merged groups. We don't want to modify
-    // any use that's outside the to group of interest as an edge of the same tv
-    // could be in other groups outside the to of the edge we're processing.
-    // Those other groups do not need to be modified as we are only interested
-    // in the merged groups to test if they can indeed be merged.
-    if (std::find(groups_to_merge.begin(), groups_to_merge.end(), edge->to) !=
-        groups_to_merge.end()) {
-      uses_to_modify.erase(
-          std::remove_if(
-              uses_to_modify.begin(),
-              uses_to_modify.end(),
-              [&](Expr* edge_val_use_expr) {
-                return !merged_group_exprs.count(edge_val_use_expr);
-              }),
-          uses_to_modify.end());
+    for (auto edge_val_use_expr : edge_tv->uses()) {
```

Comment:
Previously, we looked at all use exprs of a TV associated with an edge. I changed it to change only the exprs appearing in the consumer group of the edge. To me, this seems like simpler logic. For example, it's not necessary to do filtering like excluding use exprs that appear in the producer group.